core/src/main/scala/kafka/log/Log.scala

```diff
@@ -760,6 +768,12 @@ class Log(@volatile private var _dir: File,
       // must fall within the range of existing segment(s). If we cannot find such a segment, it means the deletion
       // of that segment was successful. In such an event, we should simply rename the .swap to .log without having to
       // do a replace with an existing segment.
+      //
+      // For case 1 (log cleaning), we may have old segments before or after the swap segment that were cleaned.
```

Comment:
I think I'm a little confused too. My understanding of this line is that we can't have offsets in the swap segment that are outside the range of the segments in log we started from. Is there a case with this new behavior that I'm missing where this isn't the case? Or is it the "existing segments" part that is confusing.

Reply:
Initially I interpreted the comment as suggesting a stronger statement about how each swapped segment related to the set of segments that it would replace. Perhaps it was really the comment that the "swap segments be composed of the original set" which confused me. 

I would suggest that we use more precise language. For example, can we say that all offsets in the swapped segment must be present in the segments that are to be replaced? It would also be helpful to clarify the following sentence:

> If we cannot find such a segment, it means the deletion of that segment was successful.

Really we are referring to a group of segments. How does the conclusion that deletion was successful follow from the fact that the swapped segment is "composed" the original segments?

Reply:
Right. Ok, I see what you are saying now. I can clarify these points. 
As for the deletion. I think the idea is that the segments that will be replaced (that "compose" the swap segment) are marked for deletion after the swap segment is created. I think the idea is that if a segment containing an offset in the range of the swap segment's offsets is missing, that means that the deletion completed. Again, I can clarify this.